sdk/core/Azure.Core/src/SequenceWriter.cs

```diff
@@ -0,0 +1,189 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Buffers;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Azure.Core
+{
+    /// <summary>
+    /// A buffer writer which writes large sequences of data into smaller shared buffers.
+    /// </summary>
+    internal sealed class SequenceWriter : IBufferWriter<byte>, IDisposable
+    {
+        private struct Buffer
+        {
+            public byte[] Array;
+            public int Written;
+        }
+
+        private Buffer[] _buffers; // this is an array so items can be accessed by ref
+        private int _count;
+        private int _bufferSize;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="SequenceWriter"/>.
+        /// </summary>
+        /// <param name="segmentSize">The size of each buffer segment.</param>
+        public SequenceWriter(int segmentSize = 4096)
```

Comment:
Where default value is coming from?

Reply:
We made it up.  @KrzysztofCwalina  is there some reasoning behind the 4096 here or should we benchmark some different values to find an optimal default?

Reply:
I think I picked it in my prototype out of a habit. It is the segment size used by Kestrel. It might be at some point worth thinking what's the right size and what pool we should be using. If I remember correctly, the ArrayPool only has 10 buffers in each size bucket, and so if we commonly allocate 10 x 4096 buffers (across all threads) we might start allocating.

Reply:
OK, I checked. The default number of arrays per bucket/partition is 32: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Buffers/SharedArrayPool.cs#L544. So, after 4096 x 32 bytes of allocations across all threads and all usages of the pool, we will start allocating. And given that I use 4096 as the habitual buffer size, it's not too far-fetched to think that 4096 is used commonly in other places in our and BCL code bases.

Reply:
We have single objects in mgmt plane that use up 39 buffers today, The [ResourceProviderData](https://github.com/Azure/azure-sdk-for-net/pull/35946/files#diff-1b59c2eae328cdd1ab7be059b1e47745d1d777df8616b68afde082b37e7939ac) object I used in the benchmarks does this.

Should we change the default to 8192?

Reply:
Worth trying, i.e. and rerunning the benchmarks

Reply:
Investigation on this is happening here https://github.com/Azure/azure-sdk-for-net/pull/38116

Reply:
Going with 16k as the default.  The other performance optimizations in https://github.com/Azure/azure-sdk-for-net/pull/38116 we will punt until after v1 ships.